pkg/ext-proc/scheduling/filter.go

```diff
@@ -0,0 +1,130 @@
+package scheduling
+
+import (
+	"fmt"
+	"math"
+
+	klog "k8s.io/klog/v2"
+
+	"ext-proc/backend"
+)
+
+type Filter interface {
+	Name() string
+	Filter(b *LLMRequest, pods []*backend.PodMetrics) ([]*backend.PodMetrics, error)
+}
+
+// filter applies current filterFunc, and then recursively applies next filters depending success or
+// failure of the current filterFunc.
+// It can be used to construct a flow chart algorithm.
+type filter struct {
+	name   string
+	filter filterFunc
+	// nextOnSuccess filter will be applied after successfully applying the current filter.
+	// The filtered results will be passed to the next filter.
+	nextOnSuccess *filter
+	// nextOnFailure filter will be applied if current filter fails.
+	// The original input will be passed to the next filter.
+	nextOnFailure *filter
+	// nextOnSuccessOrFailure is a convenience field to configure the next filter regardless of the
+	// success or failure of the current filter.
+	// NOTE: When using nextOnSuccessOrFailure, both nextOnSuccess and nextOnFailure SHOULD be nil.
+	// However if that's not the case, nextOnSuccess and nextOnFailure will be used, instead of
+	// nextOnSuccessOrFailure,  in the success and failure scenarios, respectively.
+	nextOnSuccessOrFailure *filter
+}
+
+func (f *filter) Name() string {
+	if f == nil {
+		return "nil"
+	}
+	return f.name
+}
+
+func (f *filter) Filter(b *LLMRequest, pods []*backend.PodMetrics) ([]*backend.PodMetrics, error) {
```

Comment:
This recursive model of execution does not seem necessary, we could just have a list and a for loop. I know the filters we have now are simple, but recursion is harder to profile if we ever ended up with something more complex and we needed to optimize the code.

Reply:
There are tradeoffs. With a loop we end up with many if/else branches (for the new algos to add which are more like a flow chart). My idea is to try this out on the new algos and see where it leads us. 

Definitely open for changing this as I wasn't sure this is the best model either. But so far it serves better than a list.

Reply:
I don't know if it serves better for now, but I will not nit pick for now.

Some of the nesting is moved to the configuration (e.g., https://github.com/kubernetes-sigs/llm-instance-gateway/pull/14/files#diff-c458b8dfe987c8a5018da7402eb3d77ad46123cac058555c33e496ee2ab2d4ccR13), which can get ugly when we have more than a few filters/scores.